## About the Dataset

The dataset consists of 55,000 synthetic customer transaction records from India, generated using Python’s Faker library. It contains 13 columns, each providing valuable information about customer transactions:

1. **CID (Customer ID)**: A unique identifier for each customer.
2. **TID (Transaction ID)**: A unique identifier for each transaction.
3. **Gender**: The gender of the customer (Male/Female).
4. **Age Group**: The customer’s age group, divided into categories such as 18-25, 25-45, etc.
5. **Purchase Date**: The timestamp of the transaction.
6. **Product Category**: The category of the purchased product (e.g., Electronics, Apparel, etc.).
7. **Discount Availed**: Indicates if a discount was applied (Yes/No).
8. **Discount Name**: The name of the applied discount (e.g., FESTIVE50).
9. **Discount Amount (INR)**: The amount of discount availed.
10. **Gross Amount**: The total amount before any discount.
11. **Net Amount**: The final amount after applying the discount.
12. **Purchase Method**: The method of payment used (e.g., Credit Card, Debit Card, UPI, etc.).
13. **Location**: The city where the purchase occurred.

This dataset offers a comprehensive look at customer transactions, providing insights into purchasing patterns, payment methods, and discount usage across different customer demographics.

### Initial Step: Loading Data

At the start of the project, I use the pandas library to load data from a CSV file. Pandas helps me easily work with tables, so I can analyze and clean the data:

In [1]:
import pandas as pd
import csv

To load the data, just write the path to your CSV file. Replace `data/ecommerce_dataset.csv` with the real file location on your computer:

In [3]:
data_path = 'data/ecommerce_dataset.csv'

I use `pd.read_csv()` from pandas to load the data into a DataFrame. This makes it easier to work with tables:

In [5]:
df = pd.read_csv(data_path)

### Data Analysis Before Starting Work:

The first step is to understand the dimensions of the data, specifically the number of rows and columns. I use the `.shape` method to get this information:

In [8]:
df_shape = df.shape
print(f'The dataset contains {df_shape[0]} rows and {df_shape[1]} columns.')

The dataset contains 55000 rows and 13 columns.


Next, I check the general structure of the DataFrame using the `df.info()` method, which provides details about the number of non-null values, data types, and memory usage:

In [12]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 55000 entries, 0 to 54999
Data columns (total 13 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   CID                    55000 non-null  int64  
 1   TID                    55000 non-null  int64  
 2   Gender                 55000 non-null  object 
 3   Age Group              55000 non-null  object 
 4   Purchase Date          55000 non-null  object 
 5   Product Category       55000 non-null  object 
 6   Discount Availed       55000 non-null  object 
 7   Discount Name          27415 non-null  object 
 8   Discount Amount (INR)  55000 non-null  float64
 9   Gross Amount           55000 non-null  float64
 10  Net Amount             55000 non-null  float64
 11  Purchase Method        55000 non-null  object 
 12  Location               55000 non-null  object 
dtypes: float64(3), int64(2), object(8)
memory usage: 5.5+ MB


Next, I check how much memory the dataset consumes using `.memory_usage(deep=True)`, which returns the memory usage for each column in bytes. To convert this to megabytes, I divide the result by the appropriate factor:

In [16]:
total_bytes = df.memory_usage(deep=True).sum()
total_megabytes = total_bytes / (1024 ** 2)
print(f'Total memory: {total_megabytes:.2f} MB')

Total memory: 25.70 MB


After loading the data, I use the `head()` method to preview the first few rows and the `tail()` method to preview the last three rows to ensure the data is loaded correctly:

In [20]:
df.head(3)

CID         TID  Gender     Age Group        Purchase Date  \
0  943146  5876328741  Female         25-45  30/08/2023 20:27:08   
1  180079  1018503182    Male         25-45  23/02/2024 09:33:46   
2  337580  3814082218   Other  60 and above  06/03/2022 09:09:50   

  Product Category Discount Availed    Discount Name  Discount Amount (INR)  \
0      Electronics              Yes        FESTIVE50                  64.30   
1      Electronics              Yes  SEASONALOFFER21                 175.19   
2         Clothing              Yes  SEASONALOFFER21                 211.54   

   Gross Amount   Net Amount Purchase Method   Location  
0    725.304000   661.004000     Credit Card  Ahmedabad  
1   4638.991875  4463.801875     Credit Card  Bangalore  
2   1986.372575  1774.832575     Credit Card      Delhi

In [22]:
df.tail(3)

CID         TID  Gender Age Group        Purchase Date  \
54997  649435  9587323189    Male     45-60  02/08/2024 09:30:44   
54998  982183  5032564481  Female     45-60  05/08/2020 23:57:56   
54999  126691  9172953722    Male     25-45  21/07/2022 09:05:18   

        Product Category Discount Availed    Discount Name  \
54997     Home & Kitchen              Yes         NEWYEARS   
54998  Beauty and Health              Yes  SEASONALOFFER21   
54999  Beauty and Health              Yes         NEWYEARS   

       Discount Amount (INR)  Gross Amount  Net Amount Purchase Method  \
54997                 417.63     4094.6875   3677.0575     Credit Card   
54998                 204.56     3977.4735   3772.9135     Credit Card   
54999                 380.81     3703.5819   3322.7719     Credit Card   

      Location  
54997   Mumbai  
54998  Kolkata  
54999   Mumbai

### Data Cleaning and Transformation

First, I check if there are any missing values in the dataset to ensure there are no nulls that would affect the analysis:

In [26]:
df.isnull().sum() 

CID                          0
TID                          0
Gender                       0
Age Group                    0
Purchase Date                0
Product Category             0
Discount Availed             0
Discount Name            27585
Discount Amount (INR)        0
Gross Amount                 0
Net Amount                   0
Purchase Method              0
Location                     0
dtype: int64

Before filling any missing values, I review the unique values in the `Discount Name` column to understand its content:

In [29]:
df['Discount Name'].unique()

array(['FESTIVE50', 'SEASONALOFFER21', nan, 'WELCOME5', 'SAVE10',
       'NEWYEARS'], dtype=object)

After reviewing the unique values, I fill any missing values in the `Discount Name` column with "No Discount" to ensure there are no nulls affecting the discount-related analysis:

In [32]:
df['Discount Name'] = df['Discount Name'].fillna('No Discount')

After filling the missing values in the `Discount Name` column, I check for any remaining missing values in the entire dataset to ensure the data is complete:

In [35]:
df.isnull().sum()

CID                      0
TID                      0
Gender                   0
Age Group                0
Purchase Date            0
Product Category         0
Discount Availed         0
Discount Name            0
Discount Amount (INR)    0
Gross Amount             0
Net Amount               0
Purchase Method          0
Location                 0
dtype: int64

Next, I check for duplicate rows in the dataset, as duplicates can affect the accuracy of the analysis:

In [38]:
df[df.duplicated()]

Empty DataFrame
Columns: [CID, TID, Gender, Age Group, Purchase Date, Product Category, Discount Availed, Discount Name, Discount Amount (INR), Gross Amount, Net Amount, Purchase Method, Location]
Index: []

I rename columns for better clarity and consistency:

In [41]:
df.rename(columns={'CID': 'Customer ID', 'TID': 'Transaction ID'}, inplace=True)

After renaming the columns, I check the dataset structure again using `df.info()` to verify that the column names have been updated correctly before proceeding with further data type transformations:

In [44]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 55000 entries, 0 to 54999
Data columns (total 13 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   Customer ID            55000 non-null  int64  
 1   Transaction ID         55000 non-null  int64  
 2   Gender                 55000 non-null  object 
 3   Age Group              55000 non-null  object 
 4   Purchase Date          55000 non-null  object 
 5   Product Category       55000 non-null  object 
 6   Discount Availed       55000 non-null  object 
 7   Discount Name          55000 non-null  object 
 8   Discount Amount (INR)  55000 non-null  float64
 9   Gross Amount           55000 non-null  float64
 10  Net Amount             55000 non-null  float64
 11  Purchase Method        55000 non-null  object 
 12  Location               55000 non-null  object 
dtypes: float64(3), int64(2), object(8)
memory usage: 5.5+ MB


I convert the `Purchase Date` column to the `datetime` data type to ensure proper handling of date-related operations:

In [47]:
df['Purchase Date'] = pd.to_datetime(df['Purchase Date'], dayfirst=True)

Check the data type of `Purchase Date` and preview the first few rows:

In [49]:
df['Purchase Date'].head(3)

0   2023-08-30 20:27:08
1   2024-02-23 09:33:46
2   2022-03-06 09:09:50
Name: Purchase Date, dtype: datetime64[ns]

I extract the month name from the `Purchase Date` column to make these components easily accessible for analysis:

In [53]:
df['Purchase Month Name'] = df['Purchase Date'].dt.month_name()

Check the first few rows to confirm date components are correctly extracted:

In [56]:
df[['Purchase Month Name']].head(3)

Purchase Month Name
0              August
1            February
2               March

Before changing the data types, I check the current data types of all columns using `df.dtypes` to understand the existing structure and ensure that the necessary columns are in the correct format before performing any transformations:

In [59]:
df.dtypes

Customer ID                       int64
Transaction ID                    int64
Gender                           object
Age Group                        object
Purchase Date            datetime64[ns]
Product Category                 object
Discount Availed                 object
Discount Name                    object
Discount Amount (INR)           float64
Gross Amount                    float64
Net Amount                      float64
Purchase Method                  object
Location                         object
Purchase Month Name              object
dtype: object

I convert columns such as `Gender`, `Age Group`, `Product Category`, and others to the `category` data type. This optimizes memory usage and improves performance.

In [62]:
changed_type_columns = [
    'Gender', 'Age Group', 'Product Category', 'Discount Availed', 
    'Discount Name', 'Purchase Method', 'Location', 'Purchase Month Name'
]

df[changed_type_columns] = df[changed_type_columns].astype('category')

After converting the specified columns to the `category` data type, I check the dataset again using `df.info()` to confirm that the transformations have been applied correctly. This will provide an overview of the updated data types and help ensure that the changes were successful.

In [65]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 55000 entries, 0 to 54999
Data columns (total 14 columns):
 #   Column                 Non-Null Count  Dtype         
---  ------                 --------------  -----         
 0   Customer ID            55000 non-null  int64         
 1   Transaction ID         55000 non-null  int64         
 2   Gender                 55000 non-null  category      
 3   Age Group              55000 non-null  category      
 4   Purchase Date          55000 non-null  datetime64[ns]
 5   Product Category       55000 non-null  category      
 6   Discount Availed       55000 non-null  category      
 7   Discount Name          55000 non-null  category      
 8   Discount Amount (INR)  55000 non-null  float64       
 9   Gross Amount           55000 non-null  float64       
 10  Net Amount             55000 non-null  float64       
 11  Purchase Method        55000 non-null  category      
 12  Location               55000 non-null  category      
 13  P

During the Data Cleaning and Transformation phase, I optimized memory usage by transforming data types and adding new columns. As a result, the dataset's memory usage decreased from 25.70 MB to 2.9 MB.

## Data Validation and Download for Analysis

To begin this section, I use the `describe()` function to get an overview of the numerical columns in the dataset. This helps me understand the distribution of values for `Discount Amount (INR)`, `Gross Amount`, and `Net Amount`, and check for any potential issues, such as negative values or discrepancies. To ensure clarity and precision, I set the display option to show 3 decimal places when displaying the summary statistics:

In [70]:
def describe_columns(dataframe, columns):
    result = dataframe[columns].describe()
    return result.round(3)

columns_to_describe = ['Discount Amount (INR)', 'Gross Amount', 'Net Amount']
result = describe_columns(df, columns_to_describe)
display(result)

Discount Amount (INR)  Gross Amount  Net Amount
count              55000.000     55000.000   55000.000
mean                 136.987      3012.937    2875.950
std                  165.376      1718.431    1726.128
min                    0.000       136.454    -351.120
25%                    0.000      1562.111    1429.552
50%                    0.000      2954.266    2814.911
75%                  274.115      4342.222    4211.408
max                  500.000      8394.826    8394.826

After reviewing the descriptive statistics, I notice that there are negative values in the `Net Amount` column. Therefore, I will check for any such negative values in the `Discount Amount (INR)`, `Gross Amount`, and `Net Amount` columns to identify and address any potential issues:

In [73]:
df[['Discount Amount (INR)', 'Gross Amount', 'Net Amount']].lt(0).sum()

Discount Amount (INR)      0
Gross Amount               0
Net Amount               613
dtype: int64

I noticed that there are 613 negative values in the `Net Amount` column. To ensure data integrity and avoid potential errors in the analysis, I will remove these rows with negative `Net Amount` values. This will help ensure that only valid data is used in the analysis.

In [76]:
df_cleaned = df[df['Net Amount'] >= 0].copy()

After cleaning the data by removing rows with negative values in the `Net Amount` column, I will check the summary statistics of the cleaned dataset using:

In [79]:
result_cleaned = describe_columns(df_cleaned, columns_to_describe)
display(result_cleaned)

Discount Amount (INR)  Gross Amount  Net Amount
count              54387.000     54387.000   54387.000
mean                 134.189      3043.793    2909.604
std                  163.961      1703.171    1706.287
min                    0.000       136.454       0.630
25%                    0.000      1610.986    1471.609
50%                    0.000      2983.550    2843.780
75%                  267.360      4356.490    4228.056
max                  500.000      8394.826    8394.826

I checked the data after cleaning to make sure there were no missing values or errors, and that each column had the correct data type before saving it:

In [82]:
print(df_cleaned.dtypes)

Customer ID                       int64
Transaction ID                    int64
Gender                         category
Age Group                      category
Purchase Date            datetime64[ns]
Product Category               category
Discount Availed               category
Discount Name                  category
Discount Amount (INR)           float64
Gross Amount                    float64
Net Amount                      float64
Purchase Method                category
Location                       category
Purchase Month Name            category
dtype: object


Check if there are any commas in the text columns:

In [85]:
text_cols = df_cleaned.select_dtypes(include=['object', 'category']).columns
for col in text_cols:
    comma_count = df_cleaned[col].str.contains(',', na=False).sum()
    if comma_count > 0:
        print(f"Столбец {col} содержит {comma_count} строк с запятыми")

I checked for missing valuesin the `df_cleaned` DataFrame:

In [88]:
print(df_cleaned.isna().sum())

Customer ID              0
Transaction ID           0
Gender                   0
Age Group                0
Purchase Date            0
Product Category         0
Discount Availed         0
Discount Name            0
Discount Amount (INR)    0
Gross Amount             0
Net Amount               0
Purchase Method          0
Location                 0
Purchase Month Name      0
dtype: int64


After cleaning and checking the data, I saved the final version to a CSV file. Now it can be downloaded and used for analysis or visualizations:

In [91]:
df_cleaned.to_csv(
    'cleaned_ecommerce_dataset.csv',
    index=False,
    sep=';',
    encoding='utf-8-sig',
    quoting=csv.QUOTE_NONNUMERIC,
    lineterminator='\n'
)

## Key Questions, Visualizations, and Pivot Tables for Data Analysis

### 1. Pivot Table Questions and Results

#### Goal: Understand what people buy, who buys more, and how discounts work.

* **Which product categories generate the most gross income for a specific period (year, quarter)?**

I group the data by `Product Category` and calculate the total `Gross Amount` for each year and quarter derived from `Purchase Date` to better analyze sales trends and identify seasonal fluctuations.

In [97]:
pivot_gross_amount_by_categories_years = df_cleaned.pivot_table(
    index='Product Category',
    columns=pd.to_datetime(df_cleaned['Purchase Date']).dt.year,
    values='Gross Amount',
    aggfunc='sum',
    observed=True
)

pivot_gross_amount_by_categories_years['Total Gross Amount'] = (
    pivot_gross_amount_by_categories_years.sum(axis=1)
)

pivot_gross_amount_by_categories_years_sorted = (
    pivot_gross_amount_by_categories_years.sort_values(
        by='Total Gross Amount',
        ascending=False
    )
)

pivot_gross_amount_by_categories_years_sorted.style.format(
    '{:,.0f}'
).background_gradient(
    axis=0
).set_properties(
    **{'font-weight': 'bold'},
    subset=['Total Gross Amount']
)

In [98]:
quarters_mapping = {
    'January': 'Q1',
    'February': 'Q1',
    'March': 'Q1',
    'April': 'Q2',
    'May': 'Q2',
    'June': 'Q2',
    'July': 'Q3',
    'August': 'Q3',
    'September': 'Q3',
    'October': 'Q4',
    'November': 'Q4',
    'December': 'Q4'
}

df_cleaned['Purchase Quarter'] = df_cleaned['Purchase Month Name'].map(quarters_mapping)

pivot_gross_amount_by_categories_quarter = df_cleaned.pivot_table(
    index='Product Category',
    columns='Purchase Quarter',
    values='Gross Amount',
    aggfunc='sum',
    observed=True
)

pivot_gross_amount_by_categories_quarter[
    'Total Gross Amount'
] = pivot_gross_amount_by_categories_quarter.sum(axis=1)

pivot_gross_amount_by_categories_quarter_sorted = (
    pivot_gross_amount_by_categories_quarter.sort_values(
        by='Total Gross Amount',
        ascending=False
    )
)

pivot_gross_amount_by_categories_quarter_sorted.style.format(
    '{:,.0f}'
).background_gradient(
    axis=0
).set_properties(
    **{'font-weight': 'bold'},
    subset=['Total Gross Amount']
)

**Data Insights:**

The data shows that Electronics brings in the most money, followed by Clothing and Beauty and Health. These categories consistently perform well. On the other hand, Toys & Games and Pet Care show weaker results, with noticeable drops in growth. Looking at quarterly data, Q4 is the most profitable for most categories, while sales *in Q1 and Q2 remain steady but lower.

**Recommendations:**

* Keep investing in Electronics, Clothing, and Beauty & Health as they consistently generate the most revenue.

* Investigate the decline in Electronics in 2024 to understand the cause.

* For weaker categories like Toys & Games and Pet Care, consider special offers or introducing new products to boost performance.

Focus on seasonal sales and promotions during Q4, as this quarter sees the highest customer activity.

* **Is there a correlation between the amount of discount and the total transactions by month?**

To group the data by `Purchase Month Name` and check if there's a correlation between total `Discount Amount (INR)` and number of `Transaction ID`:

In [102]:
month_order = list(quarters_mapping.keys())

pivot_discount_transactions_by_month = df_cleaned.pivot_table(
    index='Purchase Month Name',
    values=['Discount Amount (INR)', 'Transaction ID'],
    aggfunc={
        'Discount Amount (INR)': 'sum',
        'Transaction ID': 'count'
    },
    observed=True
)

pivot_discount_transactions_by_month = pivot_discount_transactions_by_month.rename(
    columns={
        'Discount Amount (INR)': 'Total Discount Amount (INR)',
        'Transaction ID': 'Total Transactions'
    }
)

pivot_discount_transactions_by_month.index = pd.CategoricalIndex(
    pivot_discount_transactions_by_month.index,
    categories=month_order,
    ordered=True
)

pivot_discount_transactions_by_month = (
    pivot_discount_transactions_by_month.sort_index()
)

pivot_discount_transactions_by_month.style.format('{:,.0f}').background_gradient(
    axis=0,
    subset=['Total Transactions']
).set_properties(
    **{'font-weight': 'bold'},
    subset=['Total Transactions']
)

In [103]:
correlation = pivot_discount_transactions_by_month[
    'Total Discount Amount (INR)'
].corr(
    pivot_discount_transactions_by_month['Total Transactions']
)

display(
    f'Correlation between total discount amount and total transactions: {correlation:.2f}'
)

'Correlation between total discount amount and total transactions: 0.89'

Thus, a correlation of 0.89 indicates a very strong positive relationship between the amount of discount and the total transactions, which may suggest that discounts do indeed encourage customers to make more purchases.

**Data Insights:**

The table shows that discounts and purchases vary every month. Discounts were highest in April and July, while the total transactions stayed mostly the same throughout the year. In months like February and September, when discounts were lower, there were also fewer purchases.

**Recommendations:**

* Increase discounts in months with fewer purchases, like February and September.

* Be aware that more discounts don't always mean more purchases — other factors like the season or marketing campaigns might also play a role.

* Track how discounts affect purchases to improve your strategy.

* **How many people use discounts?**

Group the data by `Discount Availed` and calculate the total `Net Amount` and the percentage of total `Transaction ID` where a discount was used:

In [107]:
month_order = list(quarters_mapping.keys())

pivot_discount_by_month = df_cleaned.pivot_table(
    index='Purchase Month Name',
    columns='Discount Availed',
    values=['Net Amount', 'Transaction ID'],
    aggfunc={
        'Net Amount': 'sum',
        'Transaction ID': 'count'
    },
    observed=True
)

pivot_discount_by_month.columns = [
    'Total Net Amount (No Discount)',
    'Total Net Amount (Discount Used)',
    'Total Transactions (No Discount)',
    'Total Transactions (Discount Used)'
]

pivot_discount_by_month[
    '% of Total Transactions (Discount Used)'
] = (
    pivot_discount_by_month['Total Transactions (Discount Used)'] /
    (
        pivot_discount_by_month['Total Transactions (No Discount)'] +
        pivot_discount_by_month['Total Transactions (Discount Used)']
    ) * 100
)

pivot_discount_by_month = pivot_discount_by_month[
    [
        'Total Net Amount (No Discount)',
        'Total Net Amount (Discount Used)',
        '% of Total Transactions (Discount Used)'
    ]
]

pivot_discount_by_month.index = pd.CategoricalIndex(
    pivot_discount_by_month.index,
    categories=month_order,
    ordered=True
)

pivot_discount_by_month = pivot_discount_by_month.sort_index()

pivot_discount_by_month.style.format({
    'Total Net Amount (No Discount)': '{:,.0f}',
    'Total Net Amount (Discount Used)': '{:,.0f}',
    '% of Total Transactions (Discount Used)': '{:,.2f}%'
}).background_gradient(
    axis=0,
    subset=['% of Total Transactions (Discount Used)']
).set_properties(
    **{'font-weight': 'bold'},
    subset=['% of Total Transactions (Discount Used)']
)

**Data Insights:**

Sales without discounts generate the highest revenue, but discounted products also perform strongly, especially in the later months of the year. The percentage of transactions with discounts peaks in April and December.

**Recommendations:**

* Consider offering more discounts in January and February to boost purchases during slower months.

* In peak months like April and December, think about additional discounts to increase sales.

* **Нow much total revenue was generated by each age group and gender?**

Group the data by `Age Group` and `Gender`, and calculate the count of unique `Customer ID`, the total `Net Amount`, and the average and median `Net Amount` for each combination of age group and gender:

In [112]:
pivot_net_amount_by_age_gender = df_cleaned.pivot_table(
    index=['Age Group', 'Gender'],
    values=['Customer ID', 'Net Amount'],
    aggfunc={
        'Customer ID': 'nunique',
        'Net Amount': ['mean', 'median', 'sum']
    },
    observed=True
)

pivot_net_amount_by_age_gender.columns = [
    'Unique Customers',
    'Average Net Amount',
    'Median Net Amount',
    'Total Net Amount'
]

pivot_net_amount_by_age_gender.style.format(
    {
        'Unique Customers': '{:,.0f}',
        'Average Net Amount': '{:,.2f}',
        'Median Net Amount': '{:,.2f}',
        'Total Net Amount': '{:,.0f}'
    }
).background_gradient(axis=0)

**Data Insights:**

People aged 25–45 generate the highest total revenue, with each gender group in this range contributing over 21 million INR. The average and median net amounts across all age groups stay fairly consistent, usually ranging between 2,800 and 3,000 INR, showing that customer spending per order is stable. The median values are very close to the averages, which means there aren’t many extreme or unusual purchases affecting the data. Across most age groups, users who identify as “Other” tend to spend a bit more than males and females, both on average and in terms of median spending.

**Recommendations:**

* Focus your marketing on the 25–45 age group, as they not only buy more but also bring in the highest revenue.

* Keep supporting all gender identities equally — all three groups (Female, Male, Other) show very similar spending habits.

* Consider age-based personalized offers — while younger (<18) and older (60+) customers purchase less often, their average and median spend per order is still high.

* **How do people pay, and how much do they spend with each method?**

Group the data by `Purchase Method` and calculate the total of `Net Amount`, total number of transactions `Transaction ID`, and the percentage share of each in total transactions:

In [116]:
pivot_net_amount_by_payment_method = df_cleaned.pivot_table(
    index='Purchase Method',
    values=['Net Amount', 'Transaction ID'],
    aggfunc={'Net Amount': ['mean', 'sum'], 'Transaction ID': 'count'},
    observed=True
)

pivot_net_amount_by_payment_method.columns = [
    'Average Net Amount',
    'Total Net Amount',
    'Total Transactions'
]

pivot_net_amount_by_payment_method = pivot_net_amount_by_payment_method.sort_values(
    by='Total Net Amount', ascending=False
)

pivot_net_amount_by_payment_method.style.format(
    {
        'Average Net Amount': '{:,.2f}',
        'Total Net Amount': '{:,.0f}',
        'Total Transactions': '{:,.0f}'
    }
).background_gradient(axis=0)

**Data Insights:**

Credit cards are the most popular payment method, accounting for the highest total spend — over 63 million INR.
They are followed by debit cards with around 40 million INR.
UPI services (PhonePe, Paytm, Google Pay) together contribute approximately 23 million INR, coming close to debit card usage.
Net banking and cash on delivery are used less frequently — 15.7 million INR and 7.7 million INR, respectively.
The average transaction amount across all methods remains steady, typically between 2,800–2,900 INR.

**Recommendations:**

* Keep promoting card payments, as they are the most popular.

* Focus more on UPI apps for mobile users, offering bonuses or cashback.

* Improve net banking, as it's growing.

* Encourage switching from cash on delivery to digital payments with discounts or rewards.

### Visualization Questions and Tableau Link

The `Flexible Sales and Transaction Analytics Dashboard` is designed to analyze key sales and transaction metrics with flexible time period and parameter settings. It enables the study of sales volumes, distribution by city, discount impact, and transactions by product categories and age groups, adapting to the selected filters.

#### Sales and Transaction Analytics Dashboard
[View the dashboard on Tableau](https://public.tableau.com/shared/CC5QCNQS4?:display_count=n&:origin=viz_share_link)

**1. Sales Volumes (Net and Gross Amount)**

This chart displays sales volumes (net and gross amounts) for the selected time period, set by filters (year, quarter, month, day, time of day). The X-axis shows the time period, and the Y-axis shows the amount in INR. It helps track sales trends based on the applied filters.

**2. Sales Distribution by City**

The map visualizes the distribution of sales across cities in India. The size of the dots reflects the sales volume in each city. Filters allow selecting a specific city or all cities to see the geographical distribution of sales.

**3. Discount Contribution to Net Amount**

The pie chart illustrates the contribution of discounts to the net sales amount. It breaks down sales by discount types (e.g., no discount, seasonal offers, etc.) and helps understand the share of sales attributed to different discount programs.

**4. Category Peak Transactions**

The line chart shows the average number of transactions by product categories over the selected period. Each line represents a category (e.g., Beauty & Health, Electronics). It helps identify transaction peaks by category based on the applied filters.

**5. Discounted Transaction Share by Age Groups and Product Categories (%)**

The table displays the share of discounted transactions by age groups (e.g., 18-25 and 25-45) and product categories for the selected period. It allows comparing how often discounts are applied based on customer age and product categories.

#### The filters are located on the right side of the dashboard and let you easily control which data is shown.

**1. Aggregation Level**

This filter sets the data aggregation level: year, quarter, month, day, or time of day (morning, day, evening, night). It affects all charts, determining how data is grouped by time.

**2. Year of Purchase**

Allows selecting one or multiple years for analysis. Charts update based on the selected period.

**3. Month Name**

Filter to select a specific month or all months. Helps analyze data at the month level within the selected year.

**4. Weekday Name**

Allows selecting a specific weekday (e.g., Monday, Saturday) or all days. Useful for analyzing transactions by weekday.

**5. Time Slot**

Filter to select a time of day (morning, day, evening, night). Helps understand when sales are most frequent during the day.

**6. Location**

Filter to select a specific city or all cities. Allows focusing on sales in a particular city or comparing all cities.

**7. Age Group**

Allows selecting an age group (e.g., 18-25, 25-45) . Helps analyze customer behavior based on age.

**8. Product Category**

Filter to select a discount type (e.g., SEASONALOFFER21, WELCOME) or all discounts. Helps understand the impact of specific discounts on sales.